File: sections/40_data_structures_mixed_contest/problems/k_budget_prefix_search/solution.py
import sys
def main():
    d=sys.stdin.buffer.read().split()
    if not d:return
    n,q=map(int,d[:2]);a=list(map(int,d[2:2+n]));total=[0]*(4*n);pref=[0]*(4*n)
    def pull(x):total[x]=total[2*x]+total[2*x+1];pref[x]=max(pref[2*x],total[2*x]+pref[2*x+1])
    def build(x,l,r):
        if r-l==1:total[x]=pref[x]=a[l];return
        m=(l+r)//2;build(2*x,l,m);build(2*x+1,m,r);pull(x)
    def update(x,l,r,p,v):
        if r-l==1:total[x]=pref[x]=v;return
        m=(l+r)//2
        if p<m:update(2*x,l,m,p,v)
        else:update(2*x+1,m,r,p,v)
        pull(x)
    def first(x,l,r,ql,budget,running):
        if r<=ql:return -1,running
        if ql<=l and running+pref[x]<=budget:return -1,running+total[x]
        if r-l==1:return l,running
        m=(l+r)//2;answer,running=first(2*x,l,m,ql,budget,running)
        return (answer,running) if answer>=0 else first(2*x+1,m,r,ql,budget,running)
    build(1,0,n);j=2+n;out=[]
    for _ in range(q):
        c=d[j];i=int(d[j+1])-1;x=int(d[j+2]);j+=3
        if c==b'U':update(1,0,n,i,x)
        else:answer,_=first(1,0,n,i,x,0);out.append(str(answer+1 if answer>=0 else -1))
    print('\n'.join(out))
if __name__=="__main__":main()

File: sections/40_data_structures_mixed_contest/problems/k_budget_prefix_search/solution.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;
struct Node {
  ll sum, pref;
};
struct Seg {
  int n;
  vector<Node> t;
  Seg(vector<ll> &a) : n(a.size()), t(4 * n) { build(1, 0, n, a); }
  void pull(int x) {
    t[x] = {t[2 * x].sum + t[2 * x + 1].sum,
            max(t[2 * x].pref, t[2 * x].sum + t[2 * x + 1].pref)};
  }
  void build(int x, int l, int r, vector<ll> &a) {
    if (r - l == 1) {
      t[x] = {a[l], a[l]};
      return;
    }
    int m = (l + r) / 2;
    build(2 * x, l, m, a);
    build(2 * x + 1, m, r, a);
    pull(x);
  }
  void set_value(int x, int l, int r, int p, ll v) {
    if (r - l == 1) {
      t[x] = {v, v};
      return;
    }
    int m = (l + r) / 2;
    if (p < m)
      set_value(2 * x, l, m, p, v);
    else
      set_value(2 * x + 1, m, r, p, v);
    pull(x);
  }
  int first(int x, int l, int r, int ql, ll budget, ll &sum) {
    if (r <= ql)
      return -1;
    if (ql <= l && sum + t[x].pref <= budget) {
      sum += t[x].sum;
      return -1;
    }
    if (r - l == 1)
      return l;
    int m = (l + r) / 2, res = first(2 * x, l, m, ql, budget, sum);
    return res < 0 ? first(2 * x + 1, m, r, ql, budget, sum) : res;
  }
};
int main() {
  ios::sync_with_stdio(false);
  cin.tie(nullptr);
  int n, q;
  if (!(cin >> n >> q))
    return 0;
  vector<ll> a(n);
  for (ll &x : a)
    cin >> x;
  Seg s(a);
  while (q--) {
    char c;
    int i;
    ll x;
    cin >> c >> i >> x;
    --i;
    if (c == 'U')
      s.set_value(1, 0, n, i, x);
    else {
      ll sum = 0;
      int answer = s.first(1, 0, n, i, x, sum);
      cout << (answer < 0 ? -1 : answer + 1) << '\n';
    }
  }
}
